### PR TITLE
List topics

### PR DESCRIPTION
**Pull request by joao-conde**
| - | - |
| --- | --- |
| Issue | -- |
| Dependencies | -- |
| Decisions | 1. `listTopics` method on both producer/consumer interfaces<br>2. base implements the `listTopics` as calling it on the producer or consumer<br>2. Kafka adapter `listTopics` is simply calling KafkaJS client `listTopics`|

### CODE REVIEW

File: js/kafka/kafka-client.js [hunk 1]
```diff
@@ -71,10 +71,14 @@ export class KafkaClient extends Client {
...
             }
-        });
+        }).admin();
     }
 
     get client() {
         return this._client;
     }
+
+    get topics() {
```
Comment (gcandal) [inline js/kafka/kafka-client.js]: This can't be `get` because it's `async`

File: js/kafka/kafka-client.js [hunk 2]
```diff
@@ -71,10 +71,14 @@ export class KafkaClient extends Client {
                 retries: retries,
                 maxInFlightRequests: maxInFlightRequests
             }
-        });
+        }).admin();
```
Comment (gcandal) [inline js/kafka/kafka-client.js]: Do this only when using methods that require it

**Review by gcandal — CHANGES_REQUESTED** [review]
I believe that because of our questionable dual consumer/producer design this needs a slightly different approach:
- `async listTopics` in `API`: uses either the producer or consumer, depending on what exists. If none does, ensures a producer and uses it. This producer/consumer retrieval should be its own method.
- `async listTopics` in `Producer` and `Consumer`: unimplemented
- `KafkaConsumer` and `KafkaProducer` implement `async listTopics` the same way

**Comment by joao-conde** [comment]
> 
> 
> I believe that because of our questionable dual consumer/producer design this needs a slightly different approach:
> 
>     * `async listTopics` in `API`: uses either the producer or consumer, depending on what exists. If none does, ensures a producer and uses it. This producer/consumer retrieval should be its own method.
> 
>     * `async listTopics` in `Producer` and `Consumer`: unimplemented
> 
>     * `KafkaConsumer` and `KafkaProducer` implement `async listTopics` the same way

But are we not:

a) duplicating the fetching topics code (in kafka producer and consumer)
b) conceptually making a mistake, since the available topics are for a given client connection and not to a specific producer/consumer

**Comment by gcandal** [comment]
a) Yes, but:
 - It's one line
 - They don't share an ancestor where you can place the functionality
 - Your solution exposes `Client` as an adapter, which is not true. Until now the `API` never accessed it

b) No, because:
 - "connection" doesn't exist as a high-level concept that the lib clients are exposed to
 - They share the same connection anyway

File: js/base.js [hunk 3]
```diff
@@ -89,6 +89,18 @@ export class API {
         await this.consumer.consume(topic, { callback: callback, ...options });
     }
 
+    /**
+     * Lists the available topics.
+     *
+     * @param {Function|Object} options Object that includes configuration variables.
+     * @returns {String[]} The topic names.
```
Comment (gcandal) [inline js/base.js]: In vanilla javascript we don't go so far with the typing
```suggestion
     * @returns {Array} The topic names.
```

File: js/base.js [hunk 4]
```diff
@@ -89,6 +89,18 @@ export class API {
         await this.consumer.consume(topic, { callback: callback, ...options });
     }
 
+    /**
+     * Lists the available topics.
+     *
+     * @param {Function|Object} options Object that includes configuration variables.
```
Comment (gcandal) [inline js/base.js]: Unless optionals are essential for what you need to do let's avoid them, as it brings extra complexity. It's not only an extra argument, it's writing the types for it and making sure the options aren't kafka-specific.

**Review by gcandal — CHANGES_REQUESTED** [review]
Add the Typescript definition.

**Review by joamag — APPROVED** [review]

**Comment by joamag** [comment]
@joao-conde changed method naming

**Comment by gcandal** [comment]
> @joao-conde changed method naming

@joamag we had gotten the naming from https://github.com/ripe-tech/ripe-twitch-api-js/blob/f7d17c894b5b9e1dccad80aca7dd5ea1f3277366/js/order.js#L15

This is not exactly LCRUD so it's not the same case though.

**Comment by joamag** [comment]
> > @joao-conde changed method naming
> 
> @joamag we had gotten the naming from https://github.com/ripe-tech/ripe-twitch-api-js/blob/f7d17c894b5b9e1dccad80aca7dd5ea1f3277366/js/order.js#L15
> 
> This is not exactly LCRUD so it's not the same case though.

Yeah it seemed simple to call it just topics

**Pull request merged**